File: apps/api/app/tools/reddit.py
"""Reddit discussion lookup tool."""

import re
import time
from datetime import datetime
from typing import Any

from crewai.tools import tool

from app.tools.base import CredentialError, resolve_credential_value


def get_reddit_client(credential: dict[str, Any]):
    """Get a Reddit client with credentials.
    
    Args:
        credential: Credential document with type='reddit'
        
    Returns:
        PRAW Reddit instance
        
    Raises:
        CredentialError: If credentials are missing or invalid
    """
    try:
        import praw
    except ImportError:
        raise CredentialError("praw package not installed. Run: pip install praw")
    
    client_id = resolve_credential_value(credential, "redditClientId")
    client_secret = resolve_credential_value(credential, "redditClientSecret")
    user_agent = credential.get("redditUserAgent", "SanityContentGapBot/1.0")
    
    # If user_agent is an env var reference, resolve it
    if credential.get("storageMethod") == "env" and user_agent:
        import os
        user_agent = os.environ.get(user_agent, user_agent)
    
    return praw.Reddit(
        client_id=client_id,
        client_secret=client_secret,
        user_agent=user_agent,
    )


# Subreddits relevant to CMS, web dev, and AI
RELEVANT_SUBREDDITS = [
    "webdev", "programming", "technology", "learnprogramming",
    "Wordpress", "frontend", "backend", "smallbusiness",
    "Entrepreneur", "content_marketing", "SEO", "Marketing",
    "FullStack", "nocode", "sideproject", "javascript",
    "reactjs", "nextjs", "jamstack", "headlessCMS", "cms",
    "MachineLearning", "artificial", "ChatGPT", "LocalLLaMA",
]


@tool
def reddit_discussion_lookup(query: str, credential: dict[str, Any]) -> str:
    """
    Find relevant Reddit discussions and unanswered questions about the topic.
    Returns threads and comments from relevant subreddits.
    
    Args:
        query: Topic to search for
        credential: Reddit API credential document
        
    Returns:
        Analysis of Reddit discussions including top posts and common questions
        
    Raises:
        CredentialError: If Reddit credentials are missing or invalid
    """
    reddit = get_reddit_client(credential)
    
    all_posts = []
    all_comments = []
    pattern = rf"\b{re.escape(query)}\b"
    
    # Search relevant subreddits
    for subreddit_name in RELEVANT_SUBREDDITS:
        try:
            subreddit = reddit.subreddit(subreddit_name)
            
            for submission in subreddit.search(query, limit=10):
                # Check if post is relevant
                if (re.search(pattern, submission.title, re.IGNORECASE) or
                    query.lower() in submission.title.lower() or
                    "cms" in submission.title.lower() or
                    "sanity" in submission.title.lower() or
                    "headless" in submission.title.lower() or
                    "ai" in submission.title.lower()):
                    
                    post_data = {
                        "title": submission.title,
                        "subreddit": subreddit_name,
                        "score": submission.score,
                        "num_comments": submission.num_comments,
                        "url": submission.url,
                        "permalink": f"https://reddit.com{submission.permalink}",
                        "created": datetime.fromtimestamp(submission.created_utc).strftime("%Y-%m-%d"),
                        "selftext": submission.selftext[:500] if submission.selftext else "",
                    }
                    
                    # Avoid duplicates
                    if not any(p["permalink"] == post_data["permalink"] for p in all_posts):
                        all_posts.append(post_data)
                    
                    # Get top comments
                    try:
                        submission.comments.replace_more(limit=0)
                        for comment in submission.comments.list()[:5]:
                            if hasattr(comment, "body"):
                                all_comments.append({
                                    "thread_title": submission.title,
                                    "subreddit": subreddit_name,
                                    "comment": comment.body[:300],
                                    "score": comment.score,
                                })
                    except Exception:
                        pass
            
            # Rate limiting
            time.sleep(1)
            
        except Exception:
            continue
    
    # Also search r/all for AI + CMS mentions
    try:
        for submission in reddit.subreddit("all").search(f"cms ai content {query}", limit=15):
            post_data = {
                "title": submission.title,
                "subreddit": submission.subreddit.display_name,
                "score": submission.score,
                "num_comments": submission.num_comments,
                "url": submission.url,
                "permalink": f"https://reddit.com{submission.permalink}",
                "created": datetime.fromtimestamp(submission.created_utc).strftime("%Y-%m-%d"),
                "selftext": submission.selftext[:500] if submission.selftext else "",
            }
            
            if not any(p["permalink"] == post_data["permalink"] for p in all_posts):
                all_posts.append(post_data)
    except Exception:
        pass
    
    if not all_posts:
        return f"No Reddit discussions found for '{query}'"
    
    # Sort by engagement
    all_posts.sort(key=lambda x: x["score"] + x["num_comments"], reverse=True)
    
    result = f"""
REDDIT DISCUSSIONS
==================
Search query: "{query}"
Total relevant posts found: {len(all_posts)}
Total comments collected: {len(all_comments)}

Top Discussions by Engagement:
"""
    
    for post in all_posts[:15]:
        result += f"""
Title: {post['title']}
  Subreddit: r/{post['subreddit']}
  Score: {post['score']} | Comments: {post['num_comments']} | Date: {post['created']}
  URL: {post['permalink']}
"""
        if post["selftext"]:
            preview = post["selftext"][:200].replace("\n", " ")
            result += f"  Preview: {preview}...\n"
    
    # Extract questions
    questions = [p["title"] for p in all_posts if "?" in p["title"]]
    if questions:
        result += "\nCommon Questions Asked:\n"
        for q in questions[:10]:
            result += f"  - {q}\n"
    
    # Sample comments
    if all_comments:
        result += "\nSample Comments (potential content ideas):\n"
        all_comments.sort(key=lambda x: x["score"], reverse=True)
        for comment in all_comments[:10]:
            preview = comment["comment"][:150].replace("\n", " ")
            result += f"  - [r/{comment['subreddit']}] {preview}...\n"
    
    return result

File: packages/mastra-backend/src/mastra/tools/reddit.ts
/**
 * Reddit discussion lookup tool.
 * Port of apps/api/app/tools/reddit.py
 *
 * Uses the Reddit OAuth API (same as PRAW) for reliable results.
 * Falls back to the public JSON API if no credentials are available.
 */

import { createTool } from "@mastra/core/tools";
import { z } from "zod";
import { type CredentialConfig, resolveCredentialValue } from "../../sanity.js";

const RELEVANT_SUBREDDITS = [
  "webdev", "programming", "technology", "learnprogramming",
  "Wordpress", "frontend", "backend", "smallbusiness",
  "Entrepreneur", "content_marketing", "SEO", "Marketing",
  "FullStack", "nocode", "sideproject", "javascript",
  "reactjs", "nextjs", "jamstack", "headlessCMS", "cms",
  "MachineLearning", "artificial", "ChatGPT", "LocalLLaMA",
];

interface RedditPost {
  title: string;
  subreddit: string;
  score: number;
  numComments: number;
  permalink: string;
  created: string;
  selftext: string;
}

/**
 * Strip search-engine operators the LLM may have added.
 * The query should be plain topic keywords, not a Google-style search.
 */
function sanitizeQuery(raw: string): string {
  return raw
    .replace(/site:\S+/gi, "")       // site:reddit.com etc.
    .replace(/inurl:\S+/gi, "")      // inurl:reddit
    .replace(/intitle:\S+/gi, "")    // intitle:...
    .replace(/"reddit"/gi, "")       // literal "reddit"
    .replace(/\breddit\b/gi, "")     // bare word "reddit"
    .replace(/\breddit\.com\b/gi, "") // reddit.com
    .replace(/\s{2,}/g, " ")         // collapse multiple spaces
    .trim();
}

/**
 * Get an OAuth access token from Reddit using client credentials.
 */
async function getRedditToken(clientId: string, clientSecret: string, userAgent: string): Promise<string | null> {
  try {
    const resp = await fetch("https://www.reddit.com/api/v1/access_token", {
      method: "POST",
      headers: {
        Authorization: `Basic ${Buffer.from(`${clientId}:${clientSecret}`).toString("base64")}`,
        "User-Agent": userAgent,
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: "grant_type=client_credentials",
      signal: AbortSignal.timeout(10_000),
    });

    if (!resp.ok) {
      console.warn(`[reddit] OAuth token request failed: ${resp.status}`);
      return null;
    }

    const data: any = await resp.json();
    return data.access_token ?? null;
  } catch (e: any) {
    console.warn(`[reddit] OAuth error: ${e.message}`);
    return null;
  }
}

/**
 * Search a subreddit using the OAuth API (oauth.reddit.com).
 */
async function searchSubreddit(
  sub: string,
  query: string,
  token: string,
  userAgent: string
): Promise<any[]> {
  const url = `https://oauth.reddit.com/r/${sub}/search?q=${encodeURIComponent(query)}&restrict_sr=1&limit=10&sort=relevance&t=year`;
  const resp = await fetch(url, {
    headers: {
      Authorization: `Bearer ${token}`,
      "User-Agent": userAgent,
    },
    signal: AbortSignal.timeout(10_000),
  });

  if (!resp.ok) return [];

  const data: any = await resp.json();
  return data?.data?.children ?? [];
}

/**
 * Fallback: search via the public JSON API (no auth, may be rate-limited).
 */
async function searchSubredditPublic(sub: string, query: string): Promise<any[]> {
  const url = `https://www.reddit.com/r/${sub}/search.json?q=${encodeURIComponent(query)}&restrict_sr=1&limit=5&sort=relevance&t=year`;
  const resp = await fetch(url, {
    headers: { "User-Agent": "SanityContentGapBot/1.0 (TypeScript)" },
    signal: AbortSignal.timeout(10_000),
  });

  if (!resp.ok) return [];

  const data: any = await resp.json();
  return data?.data?.children ?? [];
}

export const redditDiscussionLookup = createTool({
  id: "reddit_discussion_lookup",
  description:
    "Find relevant Reddit discussions about a topic. " +
    "Searches popular subreddits for posts and comments. " +
    "IMPORTANT: Pass only plain topic keywords (e.g. 'headless CMS comparison'). " +
    "Do NOT add 'reddit', 'site:reddit.com', or any search operators — the tool already searches Reddit.",
  inputSchema: z.object({
    query: z.string().describe("Plain topic keywords to search for (no search operators)"),
    credentialJson: z
      .string()
      .describe("JSON-encoded Reddit credential (auto-injected, do not set)"),
  }),
  execute: async ({ context: { query: rawQuery, credentialJson } }) => {
    const query = sanitizeQuery(rawQuery);
    if (!query) {
      return `Reddit search error: query was empty after sanitization (original: "${rawQuery}")`;
    }

    const allPosts: RedditPost[] = [];
    const seenPermalinks = new Set<string>();

    // Try to get OAuth credentials
    let token: string | null = null;
    let userAgent = "SanityContentGapBot/1.0 (TypeScript)";
    let useOAuth = false;

    if (credentialJson) {
      try {
        const cred: CredentialConfig = JSON.parse(credentialJson);
        const clientId = resolveCredentialValue(cred, "redditClientId");
        const clientSecret = resolveCredentialValue(cred, "redditClientSecret");
        userAgent = resolveCredentialValue(cred, "redditUserAgent") || userAgent;

        if (clientId && clientSecret) {
          token = await getRedditToken(clientId, clientSecret, userAgent);
          if (token) {
            useOAuth = true;
            console.log(`[reddit] Using OAuth API (authenticated)`);
          }
        }
      } catch {
        // Fall through to public API
      }
    }

    if (!useOAuth) {
      console.log(`[reddit] Using public JSON API (may be rate-limited)`);
    }

    // Search subreddits
    const subredditsToSearch = RELEVANT_SUBREDDITS.slice(0, useOAuth ? 15 : 8);

    for (const sub of subredditsToSearch) {
      try {
        const children = useOAuth
          ? await searchSubreddit(sub, query, token!, userAgent)
          : await searchSubredditPublic(sub, query);

        for (const child of children) {
          const d = child.data;
          if (!d?.permalink || seenPermalinks.has(d.permalink)) continue;
          seenPermalinks.add(d.permalink);

          allPosts.push({
            title: d.title ?? "",
            subreddit: d.subreddit ?? sub,
            score: d.score ?? 0,
            numComments: d.num_comments ?? 0,
            permalink: `https://reddit.com${d.permalink}`,
            created: d.created_utc
              ? new Date(d.created_utc * 1000).toISOString().slice(0, 10)
              : "unknown",
            selftext: (d.selftext ?? "").slice(0, 500),
          });
        }

        // Rate limiting — shorter delay with OAuth
        await new Promise((r) => setTimeout(r, useOAuth ? 500 : 1200));
      } catch {
        continue;
      }
    }

    // Also search r/all for broader coverage (OAuth only)
    if (useOAuth && token) {
      try {
        const allChildren = await searchSubreddit("all", query, token, userAgent);
        for (const child of allChildren) {
          const d = child.data;
          if (!d?.permalink || seenPermalinks.has(d.permalink)) continue;
          seenPermalinks.add(d.permalink);

          allPosts.push({
            title: d.title ?? "",
            subreddit: d.subreddit ?? "all",
            score: d.score ?? 0,
            numComments: d.num_comments ?? 0,
            permalink: `https://reddit.com${d.permalink}`,
            created: d.created_utc
              ? new Date(d.created_utc * 1000).toISOString().slice(0, 10)
              : "unknown",
            selftext: (d.selftext ?? "").slice(0, 500),
          });
        }
      } catch {
        // Non-fatal
      }
    }

    if (!allPosts.length) {
      return `No Reddit discussions found for '${query}'${!useOAuth ? " (note: unauthenticated — Reddit may be rate-limiting. Add Reddit credentials for better results)" : ""}`;
    }

    // Sort by engagement
    allPosts.sort((a, b) => b.score + b.numComments - (a.score + a.numComments));

    let result = `
REDDIT DISCUSSIONS
==================
Search query: "${query}"
Total relevant posts found: ${allPosts.length}
${useOAuth ? "(Authenticated via Reddit API)" : "(Public API — may have limited results)"}

Top Discussions by Engagement:
`;

    for (const post of allPosts.slice(0, 15)) {
      result += `
Title: ${post.title}
  Subreddit: r/${post.subreddit}
  Score: ${post.score} | Comments: ${post.numComments} | Date: ${post.created}
  URL: ${post.permalink}
`;
      if (post.selftext) {
        result += `  Preview: ${post.selftext.slice(0, 200).replace(/\n/g, " ")}...\n`;
      }
    }

    // Extract questions
    const questions = allPosts.filter((p) => p.title.includes("?")).map((p) => p.title);
    if (questions.length) {
      result += "\nCommon Questions Asked:\n";
      for (const q of questions.slice(0, 10)) result += `  - ${q}\n`;
    }

    return result;
  },
});
